core/src/main/java/hudson/model/Job.java

```diff
@@ -572,6 +576,13 @@ public void removeProperty(JobProperty<? super JobT> jobProp) throws IOException
      * this job.
      */
     public <T extends JobProperty> T getProperty(Class<T> clazz) {
+        if (clazz == BuildDiscarderProperty.class && logRotator != null) {
+            return clazz.cast(new BuildDiscarderProperty(logRotator));
+        }
+        return _getProperty(clazz);
+    }
+
+    private <T extends JobProperty> T _getProperty(Class<T> clazz) {
```

Comment:
Seems it doesn't fit into java naming convention: `Methods should be verbs, in mixed case with the first letter lowercase, with the first letter of each internal word capitalized.`  Maybe call `getProperty2()`?